examples/stencils.cpp

```diff
@@ -0,0 +1,237 @@
+  }
+
+  RefPtr<JS::Stencil> find(std::string const &key) {
+    std::lock_guard<std::mutex> lk(_mx);
+    auto it = _cache.find(key);
+    return it == _cache.end() ? nullptr : it->second;
+  }
+
+  void insert(std::string const &key, RefPtr<JS::Stencil> const &val) {
+    std::lock_guard<std::mutex> lk(_mx);
+    _cache.insert(std::pair{key, val});
+  }
+
+ private:
+  std::mutex _mx;
+  std::map<std::string, RefPtr<JS::Stencil>> _cache;
+};
+
+// Script compiler and executer
+class Job {
+ public:
+  Job(JSCache &cache);
+  ~Job() {
+    if (_fc) JS::DestroyFrontendContext(_fc);
+  };
+
+  void execute_script(JSContext *cx, std::string const &script,
+                      const char *filename, unsigned long linenumber);
+
+ private:
+  RefPtr<JS::Stencil> compile_script(JSContext *cx, std::string const &script,
+                                     const char *filename,
+                                     unsigned long linenumber);
+
+ private:
+  JSCache &_cache;
+  JS::FrontendContext *_fc;
+};
+
+Job::Job(JSCache &cache) : _cache(cache) {
+  static constexpr size_t kCompileStackQuota = 128 * sizeof(size_t) * 1024;
+  _fc = JS::NewFrontendContext();
+  if (_fc) JS::SetNativeStackQuota(_fc, kCompileStackQuota);
+}
+
+void Job::execute_script(JSContext *cx, std::string const &script,
+                         const char *filename, unsigned long linenumber) {
+  JS::RootedScript rscript(cx);
+  RefPtr<JS::Stencil> stencil(_cache.find(script));
+
+  if (stencil == nullptr) {
+    labeled_cout() << "Compiling script\n";
+    stencil = compile_script(cx, script, filename, linenumber);
+    if (stencil != nullptr) _cache.insert(script, stencil);
+  } else {
+    labeled_cout() << "Taking script from the cache\n";
+  }
+
+  if (stencil == nullptr) {
+    boilerplate::ReportAndClearException(cx);
+    return;
+  }
+
+  JS::InstantiateOptions instantiateOptions;
+  rscript = JS::InstantiateGlobalStencil(cx, instantiateOptions, stencil);
+  if (!rscript) {
+    boilerplate::ReportAndClearException(cx);
+    return;
+  }
+
+  JS::RootedValue val(cx);
+  if (!JS_ExecuteScript(cx, rscript, &val)) {
+    boilerplate::ReportAndClearException(cx);
+    return;
+  }
+}
+
+RefPtr<JS::Stencil> Job::compile_script(JSContext *cx,
+                                        std::string const &script,
+                                        const char *filename,
+                                        unsigned long linenumber) {
+  JS::SourceText<mozilla::Utf8Unit> source;
+  if (!_fc) return nullptr;
+
+  JS::CompileOptions opts(cx);
+  opts.setFileAndLine(filename, linenumber);
+  opts.setNonSyntacticScope(true);
+
+  if (!source.init(_fc, script.c_str(), script.size(),
+                   JS::SourceOwnership::Borrowed)) {
+    labeled_cout() << "Error initializing JS source\n";
+    ConvertFrontendErrorsToRuntimeErrors(cx, _fc, opts);
+    JS::ClearFrontendErrors(_fc);
+    return nullptr;
+  }
+
+  JS::CompilationStorage compileStorage;
```

Comment:
`JS::CompilationStorage` is removed in [bug 1857536](https://bugzilla.mozilla.org/show_bug.cgi?id=1857536).
If this branch is going to target the next ESR (128), or 125+, this variable and the parameter to `JS::CompileGlobalScriptToStencil` needs to be removed.

but if this is targeting 124 for now, I think the current code is fine.

Reply:
actually, let's remove this to make the code compatible with 125+

Reply:
Done